### PR TITLE
Autocompletion of usernames in comment boxes

### LINKED ISSUE #9755: [Discussion] Autocompletion in comment boxes
I've been looking through At.js docs https://github.com/ichord/At.js/wiki/Callbacks, https://github.com/ichord/At.js/wiki and https://github.com/publiclab/plots2/blob/main/app/assets/javascripts/atWhoAutoComplete.js. From what I understand I think the `data` key is what is supposed to hold the pre-fetched usernames

```
data: ["one", "two", "three"]
```

But instead of hardcoded names, the names are generated from recently active users on the site. I looked through the repo for the query that gets the recently active users on the site and I found this https://github.com/publiclab/plots2/blob/700fdf41a08d52bfd51d7774902dc9217ed225b5/app/controllers/users_controller.rb#L137-L142
I'm a little confused on how to display the results of that query in json format in https://github.com/publiclab/plots2/blob/700fdf41a08d52bfd51d7774902dc9217ed225b5/app/api/srch/search.rb

@jywarren @cesswairimu @RuthNjeri

### PR DESCRIPTION
**Pull request by TildaDares**
Fixes #9755

Part of larger planning issue in #9667

Test here https://unstable.publiclab.org/

https://user-images.githubusercontent.com/63427719/121555040-8ed8f900-ca0a-11eb-82fe-e58b37a24c5d.mov

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -505,6 +505,17 @@ def latest_location
     recent_locations.last
   end
 
+  def self.recently_active_users(limit = 5, order = 'last_updated DESC')
+    Rails.cache.fetch('users/active', expires_in: 24.hours) do
+      User.select('rusers.username, rusers.id, MAX(node_revisions.timestamp) AS last_updated')
+        .joins("INNER JOIN `node_revisions` ON `node_revisions`.`uid` = `rusers`.`id` ")
+        .where("node_revisions.status = 1")
```
Comment (jywarren) [inline app/models/user.rb:512]: I wonder if it would make things more efficient to add `.where("rusers.status = 1") as well here, and to the `.select()` above? We do have tens of thousands of user records. Just an idea, not sure if the join already makes this efficient!

File: app/assets/javascripts/atWhoAutoComplete.js [hunk 2]
```diff
@@ -1,8 +1,14 @@
...
 
   // settings at https://github.com/ichord/At.js/wiki/Base-Document#settings
+
+  const displayName = (item) => item.username ? item.username : item.name;
+
   var at_config = {
     at: "@",
+    displayTpl: (item) => `<li>${displayName(item)}</li>`,
+    insertTpl: (item) => `@${displayName(item)}`,
+    data: '/users/active',
```
Comment (jywarren) [inline app/assets/javascripts/atWhoAutoComplete.js:11]: How will this interact with the request down on line 15, of `$.getJSON("/api/srch/profiles?query=" + query + "&sort_by=recent&field=username", {}, function(data) {`? Does this override it?

Comment (jywarren) [reply app/assets/javascripts/atWhoAutoComplete.js:11]: Maybe even if both can co-exist, we should add a comment clarifying the difference because it's a little hard to follow.

Comment (jywarren) [reply app/assets/javascripts/atWhoAutoComplete.js:11]: aha! So `data` is the initial query after an `@`, and the `remoteFilter` is when you begin typing letters?

**Review by jywarren — CHANGES_REQUESTED** [review]
This is looking great and the video is awesome. Just had a question about the difference between 2 queries; maybe a comment would help! Thanks!!!

File: app/assets/javascripts/atWhoAutoComplete.js (hunk 2 above)
Comment (TildaDares) [reply app/assets/javascripts/atWhoAutoComplete.js:11]: @jywarren Exactly! I'll add a comment to clarify this.

File: app/models/user.rb (hunk 1 above)
Comment (TildaDares) [reply app/models/user.rb:512]: @jywarren I'll add that and see if it works.

**Review by Tlazypanda — APPROVED** [review]
@TildaDares This looks great!! :tada: another consideration can be to play around a bit with the ordering of the clauses in our query and benchmark the results :v:

**Comment by jywarren** [comment]
> @TildaDares This looks great!! 🎉 another consideration can be to play around a bit with the ordering of the clauses in our query and benchmark the results ✌️

Yes and you can also watch the rails log when running locally, and see if all the chained conditions are rolled up into a single query, and it should provide a timestamp of how long it took in development mode. But, if your local is a pretty small database, it may not really show performance issues on the scale we're looking at in the live site. 

Great ideas! Thanks!

File: app/models/user.rb [hunk 3]
```diff
@@ -505,6 +505,18 @@ def latest_location
...
 
+  def self.recently_active_users(limit = 5, order = 'last_updated DESC')
+    Rails.cache.fetch('users/active', expires_in: 24.hours) do
+      User.select('rusers.username, rusers.status, rusers.id, MAX(node_revisions.timestamp) AS last_updated')
+        .joins("INNER JOIN `node_revisions` ON `node_revisions`.`uid` = `rusers`.`id` ")
+        .where("node_revisions.status = 1")
+        .where("rusers.status = 1")
+        .group('rusers.id')
+        .order(order)
+        .limit(limit)
```
Comment (jywarren) [inline app/models/user.rb:516]: Here, i wonder -- is 5 enough? What if we changed the limit to 15? Or, alternatively, what if there were a way to narrow this list based on something specific to the current user? But, that might increase the complexity of the query and therefore slow it down a lot. 

Finally, i wonder if we could add caching to this query. That could make this really speedy and it seems OK for there to be at least an hour long cache here, maybe? Look at this example - 

https://github.com/publiclab/plots2/blob/cfba9ef309198ca0731f9c0d54214816c8915278/app/controllers/stats_controller.rb#L5

This can be added in followup, so I'll go ahead and merge this now, but these are maybe some more refinements we can make -- and they're simple enough that maybe they could be an FTO too!

**Review by jywarren — APPROVED** [review]

**Comment by jywarren** [comment]
Just noting that I think this is ready to go -- we can test it in stable, and we won't be publishing to the live site for at least a few days or maybe longer -- and the extra ideas I suggested above could be done in another PR!

Thanks, everyone! @TildaDares this is great work!

File: app/models/user.rb (hunk 3 above)
Comment (TildaDares) [reply app/models/user.rb:516]: > What if we changed the limit to 15?

@jywarren Setting the limit to 15 sounds like a good idea. 


> Finally, i wonder if we could add caching to this query. That could make this really speedy and it seems OK for there to be at least an hour long cache here, maybe? Look at this example -

The query already has a cache or is there somewhere else I'm supposed to add it? The cache has an expiration time of 24 hours so I'll change it to an hour.

https://github.com/publiclab/plots2/blob/c76e4859dcb69090ccf31ed8fe6140e4eaa567d2/app/models/user.rb#L509

**Comment by jywarren** [comment]
ah, ok, perfect! Thank you!
